packages/network/src/logic/node/ProxyStreamConnectionManager.ts

```diff
@@ -0,0 +1,211 @@
+    trackerManager: TrackerManager,
+    streamManager: StreamManager,
+    nodeToNode: NodeToNode,
+    node: Node,
+    nodeConnectTimeout: number,
+    acceptProxyConnections: boolean
+}
+
+enum State {
+    CONNECTING,
+    CONNECTED,
+    RECONNECTING
+}
+
+interface ProxyConnection {
+    state?: State,
+    reconnectionTimer?: NodeJS.Timeout
+}
+
+const DEFAULT_RECONNECTION_TIMEOUT = 10 * 1000
+
+export class ProxyStreamConnectionManager {
+    private readonly trackerManager: TrackerManager
+    private readonly streamManager: StreamManager
+    private readonly nodeToNode: NodeToNode
+    private readonly node: Node
+    private readonly nodeConnectTimeout: number
+    private readonly acceptProxyConnections: boolean
+    private readonly connections: Map<SPIDKey, Map<NodeId, ProxyConnection>>
+
+    constructor(opts: ProxyStreamConnectionManagerOptions) {
+        this.trackerManager = opts.trackerManager
+        this.streamManager = opts.streamManager
+        this.nodeToNode = opts.nodeToNode
+        this.node = opts.node
+        this.nodeConnectTimeout = opts.nodeConnectTimeout
+        this.acceptProxyConnections = opts.acceptProxyConnections
+        this.connections = new Map()
+    }
+
+    addConnection(spid: SPID, nodeId: NodeId): void {
+        if (!this.connections.has(spid.key)) {
+            this.connections.set(spid.key, new Map())
+        }
+        this.connections.get(spid.key)!.set(nodeId, {
+            state: State.CONNECTING
+        })
+    }
+
+    removeConnection(spid: SPID, nodeId: NodeId): void {
+        if (this.connections.has(spid.key)) {
+            this.connections.get(spid.key)!.delete(nodeId)
+            if ([...this.connections.get(spid.key)!].length === 0) {
+                this.connections.delete(spid.key)
+            }
+        }
+
+        this.streamManager.removeNodeFromStream(spid, nodeId)
+        if (this.streamManager.isSetUp(spid)
+            && this.streamManager.getAllNodesForStream(spid).length === 0
+            && !this.connections.has(spid.key)
+            && this.streamManager.isBehindProxy(spid)
+        ) {
+            this.streamManager.removeStream(spid)
+        }
+    }
+
+    private hasConnection(nodeId: NodeId, spid: SPID): boolean {
+        return this.connections.get(spid.key)!.has(nodeId)
+    }
+
+    private getConnection(nodeId: NodeId, spid: SPID): ProxyConnection | undefined {
+        return this.connections.get(spid.key)!.get(nodeId)!
+    }
+
+    async openOutgoingStreamConnection(spid: SPID, targetNodeId: string): Promise<void> {
+        const trackerId = this.trackerManager.getTrackerId(spid)
+        const trackerAddress = this.trackerManager.getTrackerAddress(spid)
+        try {
+            if (!this.streamManager.isSetUp(spid)) {
+                this.streamManager.setUpStream(spid, true)
+            } else if (this.streamManager.isSetUp(spid) && !this.streamManager.isBehindProxy(spid)) {
```

Comment:
As we check `!this.streamManager.isSetUp(spid)`at line 96, is there need to recheck the opposite case at line 98, 103 and 108? 

Also the method could be easier to read if we'd move lines 113-116 after line 97, and then remove the redundant `return` statements.

Reply:
Line 96 there to create a stream if necessary, it is not use specifically to enter lines 113-116 expect when a stream does not exist. Rest of the time when the function is called (if the stream already exists) it will have to pass the cases on lines 98, 103 and 108 before starting to create the proxy connection